### PR TITLE
ARROW-820: [C++] Build dependencies for Parquet library without arrow…

### PR DESCRIPTION
**Pull request by majetideepak**
… support

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -889,36 +896,50 @@ add_subdirectory(src/arrow/ipc)
```
Comment (majetideepak) [inline cpp/CMakeLists.txt]: It is okay to install the headers for now. Parquet `FindArrow.cmake` looks for `arrow/arrow.h`

**Comment by wesm** [comment]
I don't agree with this approach because I don't think it's going to be maintainable (and having Parquet-specific details in Arrow's build system doesn't seem like a good idea). I'll take a closer look tomorrow and propose an alternative solution that avoids the extra thirdparty dependencies.

**Comment by majetideepak** [comment]
Sounds good. Thanks.

**Comment by wesm** [comment]
We *just* consolidated this code in https://github.com/apache/arrow/commit/d1a9aff2937efe54fe3a5c80f7fbe19851cb71f3  I'm skeptical whether having all third parties have to link to lots of different shared libraries is the simplest solution -- you can always statically link the symbols that you need from libarrow.a, and the build time is short.

**Comment by majetideepak** [comment]
Ok.  I will just use the existing `ARROW_IPC` flag and avoid building the third party (`flatbuffers, rapidjson, hdfs`) dependencies if it is OFF. Is this reasonable?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -904,21 +913,30 @@ set(ARROW_SRCS
...
-
   src/arrow/util/bit-util.cc
   src/arrow/util/decimal.cc
 )
+if (ARROW_HDFS)
+  set(ARROW_SRCS ${ARROW_SRCS}
+    src/arrow/io/hdfs.cc
+    src/arrow/io/hdfs-internal.cc
+  )
+endif()
```
Comment (wesm) [inline cpp/CMakeLists.txt]: Is compiling this problematic? There are no additional thirdparty requirements, nor anything additionally linked.

Comment (majetideepak) [inline cpp/CMakeLists.txt]: both of them include the header `hdfs.h` located in `${HADOOP_HOME}`. The `ARROW_HDFS` flag if set to `OFF` will not include this header. These files should not be compiled then.

Comment (wesm) [inline cpp/CMakeLists.txt]: Is this a problem, though? hdfs.h is in https://github.com/apache/arrow/tree/master/cpp/thirdparty/hadoop/include

Comment (majetideepak) [inline cpp/CMakeLists.txt]: It won't be a problem if the header already exists. Will revert it.

**Review by wesm — APPROVED** [review]
+1, thanks @majetideepak!

**Review by xhochy — APPROVED** [review]
+1, this is also something more "picky" Linux distributions like Gentoo like to see :)

**Pull request closed without merging**